Documentation/Proposals/NNNN-third-party-issues.md

```diff
@@ -0,0 +1,108 @@
+# Improve robustness of recording Issues
+
+* Proposal: [SWT-NNNN](NNNN-third-party-issues.md)
+* Authors: [Rachel Brindle](https://github.com/younata), [Jonathan Grynspan](https://github.com/grynspan)
+* Status: **Awaiting review**
+* Bug: [apple/swift-testing#490](https://github.com/apple/swift-testing/issues/490)
+* Implementation: [apple/swift-testing#513](https://github.com/apple/swift-testing/pull/513)
+* Review: [To be added]
+
+## Introduction
+
+Integration with third party tools is important for the success of Swift Testing
+and the ecosystem as a whole. To support this, Swift Testing should offer tools
+more control over how custom issues are recorded and what is shown.
+
+## Motivation
+
+There are several third-party assertion libraries that developers are used to
+using, and Swift Testing should make it as easy as possible for the developers
+and maintainers of these third-party assertion libraries to integrate with
+Swift Testing.
+
+Swift Testing currently offers the `Issue.record` family of static methods,
+which provides an acceptable interface for one-off/custom assertions without
+using the `#expect` or `#require` macros. Unfortunately, the public
+`Issue.record` static method does not offer a way to precisely control the
+error information shown to the user - all Issues are recorded as "unconditional
+failure", which is false if the Issue actually happened as the result of an
+assertion failing.
+
+## Proposed solution
+
+We create a new public `Issue.record` static method, specifically to record and
+display only the information given to it:
+
+```swift
+extension Issue {
+  @discardableResult public static func record(
+    _ comment: Comment? = nil,
+    context toolContext: some Issue.Kind.ToolContext,
+    sourceLocation: SourceLocation = #_sourceLocation
+  ) -> Self
+}
+```
+
+This method is then used by tools to provide a comment of what happened, as well
+as information for the user to know what tool actually recorded the issue.
+
+## Detailed design
+
+`Issue.record` creates an issue of kind `Issue.Kind.recordedByTool`. This is a
```

Comment:
By the way, this is an opportunity for you to refine the names of the new symbols we're introducing here. If you hate `ToolContext` or `recordedByTool` or what-have-you, let's discuss.

(That's not to say you _must_ change names, just that the names we've used in my draft PR don't need to be the final ones if you think we can improve on them.)

Reply:
Good point, and I didn't do that because I was more focused on just writing the pitch.

I'll do some of my thinking on that here, and later refine I can that in to the pitch doc:

`ToolContext` isn't bad, but after more thought, I think that `Metadata` is a better suffix than `Context`. Regardless, I think that `ToolMetadata` is a better description of what's expected.

As for the Tool Prefix, I'm pretty meh about it. I wonder if maybe dropping the prefix entirely is better? `Issue.Kind.Metadata` does a decent-enough job of namespacing. It is a bit weird that a type named "Metadata" is only intended to be used by third party tools though.

Similarly, I wonder if the actual contents of `ToolContext` should have a different. Perhaps also, to better express the intent that this is metadata, we should rename the `toolName` property to something like `toolDescription`? That gives the indication that this is meant more for human-readable display (under the assumption that developers have internalized -description properties like `description` and `debugDescription` are dynamically generated for human consumption?). Which also helps make it obvious that they can also have their `ToolContext` type contain other properties if they want. Additionally, the documentation for `ToolContext` should also explicitly state that it the `ToolContext` will be included in the json stream from Swift Testing.

As for the Issue.Kind case, I still don't have a decent suggestion for something better. `recordedByTool` isn't bad, per-se, but it describes where the issue came from, not really what kind of issue it is (this might be a distinction without a difference). More thought on this is required.

I don't hate your suggestions, and I think they're really useful for helping to move the conversation along!

Reply:
`Metadata` feels like it's vague enough that we'd come regret it later, even though it's nested inside `Issue.Kind`. What about a top-level type, `ToolMetadata`, that isn't specifically bound to `Issue.Kind`?

If we _do_ grab the tool name from its module, then we don't need `toolName` at all, FWIW. I don't feel strongly about renaming it.

> As for the Issue.Kind case, I still don't have a decent suggestion for something better. recordedByTool isn't bad, per-se, but it describes where the issue came from, not really what kind of issue it is (this might be a distinction without a difference). More thought on this is required.

The hard part here is that Swift Testing itself hasn't got a clue what kind of issue something is. The existing cases of `Issue.Kind` are certainly descriptive, but in terms of the Swift Testing API that was ultimately used to produce them. For example, `#expect` produces `.expectationFailed`, and `confirmation {}` produces `.confirmationMiscounted`. So `toolAPI()` should produce `toolSomethinged()`. From a test author's point of view, it's mostly moot because they don't typically see an instance of `Issue` directly, they just see the output in CLI, Xcode, or VSCode.